ChatSecure/Classes/Model/Yap Storage/OTRXMPPRoomOccupant.swift

```diff
@@ -64,11 +65,25 @@ open class OTRXMPPRoomOccupant: OTRYapDatabaseObject, YapDatabaseRelationshipNod
     @objc open var affiliation:RoomOccupantAffiliation = .none
 
     /**When given by the server we get the room participants reall JID*/
-    @objc open var realJID:String?
+    @objc open var realJID:String? {
+        didSet {
+            if let realJid = self.realJID, let roomUniqueId = self.roomUniqueId {
+                OTRDatabaseManager.shared.readOnlyDatabaseConnection?.asyncRead({ (transaction) in
+                    if let room = OTRXMPPRoom.fetchObject(withUniqueID: roomUniqueId, transaction: transaction), let accountUniqueId = room.accountUniqueId {
+                        self.realBuddy = OTRBuddy.fetch(withUsername: realJid, withAccountUniqueId: accountUniqueId, transaction: transaction)
+                    }
+                })
+            }
+        }
+    }
+    @objc open var realBuddy:OTRBuddy?
     
     @objc open var roomUniqueId:String?
     
     @objc open func avatarImage() -> UIImage {
+        if let buddy = self.realBuddy {
+            return buddy.avatarImage
+        }
```

Comment:
You can this logic elsewhere (see below)

Reply:
I still think having an implicit database transaction here is bad news. If you accidentally have a transaction inside another transaction you'll likely deadlock. Either have a function where you pass in a transaction, or handle the logic to determine which avatar to return at a higher level.

Reply:
Yeah, I agree, I realized late yesterday that's probably what you meant, but didn't have time to fix it. Done now! Spent most of the day trying to get things to build, after Swift 4 update. I could not find ZXingObjC 3.2.2 (maybe need to update something?) but figured out I could use your xcode9 branch for now.

Reply:
You'll need to update the Carthage dependencies:

```
$ rm -rf ~/Library/Caches/org.carthage.CarthageKit
$ carthage update --platform ios --cache-builds
```